### ISSUE TITLE
Using Kubernetes to orchestrate arbitrary infrastructure 

### ISSUE THREAD

**Issue by vvcb** [lsc-sde/lsc-sde#1]
@qcaas-nhs-sjt , your comment about k8s being an orchestration platform and containers are just one of many things it can orchestrate raises all kinds of interesting possibilities.

I have been looking at https://kubevirt.io/ as an example of using k8s to control VMs for instance.

When combined with JupyterHub's architecture, this may be one way to deliver:

a. A secure VDI wrapper solution to JupyterHub's browser-based tools as discussed here - https://discourse.jupyter.org/t/run-linux-desktop-apps-in-mybinder-org-your-jupyterhub/11964/11
b. Provision Windows VMs on-demand using JupyterHub as described here by @manics - https://github.com/manics/jupyterhub-virtualdesktopspawner and https://github.com/manics/jupyter-guacamole

If this proves to be a good solution, it can potentially be a replacement for AVD and we can provide access to any private resources (e.g OHDSI, DataSHIELD, etc.) through this.

Can we please add this to our roadmap and discuss when we meet next?

It will be good to work with @manics who knows JupyterHub really well and is part of the UK society for research software engineering.

@m1p1h , if you are not already a member you should sign up - https://society-rse.org/

**Comment by vvcb** [comment]
This looks very interesting. 

We are using [JupyterHub's Z2JH](https://z2jh.jupyter.org/) at the core of our analytics environment and we have battle-tested it at LTH for a few years now.  

From a user experience POV, it will be good to keep all the functionality of Z2JH (see [architecture](https://z2jh.jupyter.org/en/stable/administrator/architecture.html)) and use a modified [KubeSpawner](https://github.com/jupyterhub/kubespawner) to spawn arbitrary resources - example of provisioning EC2 instance [here](https://github.com/manics/jupyterhub-virtualdesktopspawner/blob/main/virtualdesktopspawner/kubeec2spawner.py). KubeSpawner in turn uses the kubernetes API anyway.

@manics, it will be good to get your thoughts on this.

**Comment by manics** [comment]
This sounds cool, and should be doable! I'm very keen to see a production Kubernetes hosted TRE, especially one that can provide Windows desktops. The idea of JupyterHub as a generic resource manager has come up in HPC discussions before. Happy to chat if you want!

I've tried KubeVirt before but couldn't get it to work. I thought it was an abstraction over public cloud APIs for managing VMs, but it actually uses KVM so if you're running it on the cloud you end up with a nested VM.

Crossplane is worth a look if you haven't seen it before https://docs.crossplane.io/latest/getting-started/introduction/
I discoveted it over the holidays, it provides K8s custom resource definitions for many public cloud resources- superficially souds similar to the custom operators @qcaas-nhs-sjt mentioned except that someone else has done the work :smile:
On the plus side it's an official CNCF project https://www.cncf.io/projects/crossplane/, on the minus side it's in incubation so there's not yet a guarantee that it'll exist long term. It seems to be an autogenerated wrapper around some the cloud Terraform providers, so has pretty comprehensive coverage of the main cloud resources.

The big advantage of using CRDs is that KubeSpawner needs to be extended to handle arbitrary K8s manifests with some templating, whereas procedural code to bring up an instance and configure it's networking has to be rewritten for every cloud and resource type. Perhaps there's even a case to be made for making KubeSpawner use K8s manifest templates only even for standard pod and volume definitions, so it almost becomes "HelmSpawner" but with Jinja2 style templating?

**Comment by qcaas-nhs-sjt** [comment]
@manics good catch, I was not aware of crossplane, the concept looks relatively sound, though my only worry is that they use a similar solution in terraform for talking to the azurerm API's and despite this terraform has difficulties keeping up with all the new functionalities & parameters. As crossplane is trying to be all things to all people it will require a lot of maintenance from the community to keep on top the changes. Still that said I think a proof of concept would be valuable in this instance.

We also need to consider the security aspects of such a controller, if not applied carefully, such a controller could be misused to create exploits which could leave us exposed. We cannot rely on the IAM security of azure alone to manage this because of the level of which it would run. For example it is likely that our solution will have to provision a virtual machine, and extensions to configure tools or even just to allow the virtual machine to join an active directory domain, etc. IAM will only know that it is creating an extension, it will not really care about what that extension is doing. If such an extension were to be defined that would install malicious code onto existing servers then all of the information on these servers could be transferred out through a back door. We would therefore need to create a very strict set of IAM controls, azure policies, and resource based kubernetes RBAC. We'd also likely want to split out the workers for this from the rest of the cluster so that they would have to run on their own nodes with their own special sets of permissions so that other resources couldn't access these.

On the flip side our own custom controllers would allow us to be a lot more targetted about which resource types we were offering up and to focus in on the functionality that we're actually using which would be a lot less up-keep and wouldn't then be so reliant on this existing community. Regarding the time taken to write such controllers, I'm not overly worried I've written many such controllers in the past and once you've got a working template it doesn't tend to take much to build them, so the main part of it will be building up the base template. From a security aspect, as the custom controllers would only be able to do the things we've programmed the risk is much lower. (Though obviously we'll still need appropriate RBAC)

Regardless of whether we can use crossplane or have to write our own controllers I think that the truly time consuming part will be to define a truly secure environment and all the components & extensions that are needed to make this work properly *and* to get the security elements in place that allow us to provision it with minimal access rights. There are many different moving parts that need to be considered, not least of all the challenges of building such a thing inside of an NHS environment which has unique challenges

**Comment by manics** [comment]
All good points! I'm approaching this from the perspective of wanting the flexibility to extend a TRE to encompass arbitrary approved cloud services, where it's helpful to not need to deploy code for every change. Regardless, if all cloud resources are managed via CRDs the changes to KubeSpawner would be the same.

**Comment by qcaas-nhs-sjt** [comment]
Linking to lsc-sde/lsc-sde#37

**Comment by vvcb** [comment]
@m1p1h bringing this discussion to your attention. Especially the discussion regarding using Kubernetes to orchestrate everything vs doing this through JupyterHub and KubeSpawner. 

https://github.com/kubevirt/kubevirt looks really interesting. A full VM-based environment will be useful for some of the more complex use cases that require tools that are not easily containerised (yet).

Related to #57.

@qcaas-nhs-sjt shall we meet with @manics for a brainstorming session and see if we can come up with a design that we could possibly work together on.

**Comment by qcaas-nhs-sjt** [comment]
> @m1p1h bringing this discussion to your attention. Especially the discussion regarding using Kubernetes to orchestrate everything vs doing this through JupyterHub and KubeSpawner.
> 
> https://github.com/kubevirt/kubevirt looks really interesting. A full VM-based environment will be useful for some of the more complex use cases that require tools that are not easily containerised (yet).
> 
> Related to #57.
> 
> @qcaas-nhs-sjt shall we meet with @manics for a brainstorming session and see if we can come up with a design that we could possibly work together on.

Yeah I think that would be useful, however I'm not convinced that kubevirt is the right solution, as I understand it, it uses nested virtualization to provision the virtual machines, this adds many complexities I think can be avoided and it means that network access control becomes very difficult, not to mention that a large portion of the azure estate does not allow nested virtualization.

The cross-plane idea might be a better one, or even just building custom operators for each aspect of the system. This would allow us to not only create the VM but the network it sits upon within the azure/aws/gcp estate, we would have custom operators for each provider to convert our workspace definitions into the relevant resources in crossplane to facilitate the system.

In my mind this would work in a similar way to azure's TRE, so the workspace definition would be extended to include details needed to spin up a virtual network and virtual machines. The operator would detect this and convert this definition into the CRD definitions in crossplane, crossplane would implement these.

```mermaid
flowchart TB
AWMS[Workspace Management Portal] --> |1. Creates workspace |API((k8s API)) 
--> |2. Triggers Operator Action| AWMSO[AWMS-to-Crossplane-Azure Operator]  --> |3. Writes converted crossplane definition| API 
--> |4. Triggers Operator Action| Crossplane[Crossplane Operator]
--> |5. Creates/updates the resources based on the crossplane definition| Azure[Azure Graph API]
Crossplane --> |6. Updates status of crossplane resource|API
```

As this is a microservice architecture you could swap out the Azure AWMS-to-crossplane operator for another operator for another cloud provider such as AWS:

```mermaid
flowchart TB
AWMS[Workspace Management Portal] --> |1. Creates workspace |API((k8s API)) 
--> |2. Triggers Operator Action| AWMSO[AWMS-to-Crossplane-AWS Operator]  --> |3. Writes converted crossplane definition| API 
--> |4. Triggers Operator Action| Crossplane[Crossplane Operator]
--> |5. Creates/updates the resources based on the crossplane definition| AWS[AWS API]
Crossplane --> |6. Updates status of crossplane resource|API
```
or GCP:

```mermaid
flowchart TB
AWMS[Workspace Management Portal] --> |1. Creates workspace |API((k8s API)) 
--> |2. Triggers Operator Action| AWMSO[AWMS-to-Crossplane-GCP Operator]  --> |3. Writes converted crossplane definition| API 
--> |4. Triggers Operator Action| Crossplane[Crossplane Operator]
--> |5. Creates/updates the resources based on the crossplane definition| GCP[GCP API]
Crossplane --> |6. Updates status of crossplane resource|API
```

or even kubevirt:

```mermaid
flowchart TB
AWMS[Workspace Management Portal] --> |1. Creates workspace |API((k8s API)) 
--> |2. Triggers Operator Action| AWMSO[AWMS-to-KubeVirt Operator]  --> |3. Writes converted crossplane definition| API 
--> |4. Triggers Operator Action| KubeVirt[KubeVirt Operator] --> |5. Updates status of crossplane resource|API
```

This of course all assumes that CrossPlane works how we need it to, I would recommend that we do a proof-of-concept to prove that this works in the way we would like. Worst case scenario we cut out the middleman and have our operators talk directly to azure/aws/gcp

```mermaid

flowchart TB
AWMS[Workspace Management Portal] --> |1. Creates workspace |API((k8s API)) 
--> |2. Triggers Operator Action| AWMSO[Azure AWMS Operator]  --> |3. Creates required resources in azure| Azure[Azure Graph API]
AWMSO --> |4. Updates status of crossplane resource|API
```

**Comment by manics** [comment]
Happy to chat! I can't do Fridays and I've got limited availability this/next week but should be able to fit something in.

My thoughts on Kubevirt are similar to @qcaas-nhs-sjt
though I haven't looked at it recently. 

It could be worth looking at https://github.com/SwissDataScienceCenter/amalthea for ideas. It's a standalone project to manage jupyter-servers using CRDs.

**Comment by qcaas-nhs-sjt** [comment]
@manics sounds good

Please could you drop me an email on shaun.turner1@nhs.net with your availability and we will try to get something setup

Many thanks
